src/EditorFeatures/Core/Implementation/ExtractMethod/AbstractExtractMethodCommandHandler.cs

```diff
@@ -91,7 +91,7 @@ public bool ExecuteCommand(ExtractMethodCommandArgs args, CommandExecutionContex
                 _renameService.ActiveSession.Commit();
             }
 
-            using (context.OperationContext.AddScope(allowCancellation: true, EditorFeaturesResources.Applying_Extract_Method_refactoring))
+            using (context.OperationContext.WaitForDocumentAndAddScope(document, allowCancellation: true, EditorFeaturesResources.Applying_Extract_Method_refactoring))
             {
                 return Execute(args.SubjectBuffer, args.TextView, context.OperationContext);
             }
```

Comment:
this is a bit strange.  To operate properly, you would *not* be able to use this document anymore.  So someone passing the document along will end up getting bogus results.

It seems like teh more appropriate way to design this API would be to have this flow:

1. You start with something like a TextBuffer.
2. From that, you ask for the Workspace associated with it.
3. You try to wait on that Workspace being fully loaded.
4. you *then* attempt to get the document for that buffer, which now poitns to the current up to date solution.

All of this can be encapsulated behind a safe single call.  Then, you can be sure that the Workspace is fully loaded, you have a document associated with the buffer, and the document represents the latest solution, and not some point in the past.

Reply:
so. you suggesting GetLatestDoucument that accepts OperationContext? sure. we can do that.

Reply:
I'm not certain what GetLatestDocument is.  But i think what i would prefer the most would be an API like this:

```c#
public static Document GetOpenDocumentInCurrentContextWithChanges(
    this SourceText text, bool waitForFullyLoadedWorkspace, CancellationToken ct)
```

Basically, the same entrypoint that people can already depend on today, but with a way to say that you're willing to wait on the VSWorkspace being fully loaded.  It's a natural sibling to what we already have.

Reply:
we could have that, but then people has to write adding scope and setting the description and etc. I want to just give them 1 code they can just do. also, I dont want nested scope since it shows 2 wait dialogbox in succession.

Reply:
experience wise, what I have is what we want. 

1. operation inovked by users. 
2. we wait 2 seconds (wait dialog box thing basically), 
3. show standard title for partial load mode, then it is done, 
4. same wait dialog box stay up with different title for extract method and 
5. it removed when operation is actually done

Reply:
You can have an extension on the context like so:  

`var document = context.GetDocumentInFullyLoadedWorkspace(sourceText);`

Right?

This would do the load, having put in the information in the wait dialog indicating that we're waiting on solution load).  Then, this pops back out with the document.  This has several benfits:

1. only one Document instance you need to concernt youself with.
2. that extension method can properly create the nested scope and set teh description.  
3. that extension method would not leak out cancellation tokens.
4. like GetOpenDocumentInCurrentContextWithChanges this would just return null if it failed.  callers just have to check for null and return as they cannot proceed. (note: you can name this TryGet if you want to indicate null returning).
5. When the inner context scope pops out, the dialog goes back to whatever description the caller asked for.

In effect, instead of having teh caller need to care about this stuff (and juggle scopes), this is all encapsulated in the GetDocumentInFullyLoadedWorkspace.  Furthermore, you never have multiple different Solution/Project/Doducment snapshots that have to be considered. 

Finally, this best matches the sort of entrypoint a feature woudl want.  They're already saying "give me the doc for a buffer".  Now they simply can say "give me doc for thsi buffer in a fully loaded VS workspace".

Reply:
I didn't go that route since it created 2 scopes. since first scope is happening in GetDocumetn... and then scope finishes and then create new scope for extract method that make irritating 2 wait dialogboxes.

Reply:
> and then scope finishes and then create new scope for extract method that make irritating 2 wait dialogboxes.


hrmmm... that seems like a bug with the editor (i actually may have opened one on them about this).

To workaround that beahvior, this seems fine:

```c#
using (context.OperationContext.AddScope(allowCancellation: true, EditorFeaturesResources.Applying_Extract_Method_refactoring))
{
     var doc = context.GetDocumentInFullyLoadedWorkspace(args.SubjectBuffer.CurrentSnapshot);
     
    // rest of the work
}
```